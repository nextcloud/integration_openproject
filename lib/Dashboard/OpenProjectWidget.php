<?php

/**
 * SPDX-FileCopyrightText: 2022-2025 Jankari Tech Pvt. Ltd.
 * SPDX-FileCopyrightText: 2021-2022 Nextcloud GmbH and Nextcloud contributors
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

namespace OCA\OpenProject\Dashboard;

use OCA\OpenProject\AppInfo\Application;
use OCA\OpenProject\Service\OpenProjectAPIService;
use OCP\AppFramework\Services\IInitialState;
use OCP\Dashboard\IWidget;
use OCP\IConfig;
use OCP\IL10N;
use OCP\IURLGenerator;
use OCP\IUser;
use OCP\IUserSession;

use OCP\Util;

class OpenProjectWidget implements IWidget {

	/**
	 * @var IL10N
	 */
	private $l10n;
	/**
	 * @var IURLGenerator
	 */
	private $url;
	/**
	 * @var IInitialState
	 */
	private $initialStateService;
	/**
	 * @var IConfig
	 */
	private $config;

	/**
	 * @var IUser
	 */
	private $user;

	/**
	 * @var OpenProjectAPIService
	 */
	private OpenProjectAPIService $openProjectAPIService;


	public function __construct(
		IL10N $l10n,
		IInitialState $initialStateService,
		IURLGenerator $url,
		IConfig $config,
		IUserSession $userSession,
		OpenProjectAPIService $openProjectAPIService
	) {
		$this->initialStateService = $initialStateService;
		$this->l10n = $l10n;
		$this->url = $url;
		$this->config = $config;
		$this->user = $userSession->getUser();
		$this->openProjectAPIService = $openProjectAPIService;
	}

	/**
	 * @inheritDoc
	 */
	public function getId(): string {
		return 'openproject_notifications';
	}

	/**
	 * @inheritDoc
	 */
	public function getTitle(): string {
		return $this->l10n->t('OpenProject');
	}

	/**
	 * @inheritDoc
	 */
	public function getOrder(): int {
		return 10;
	}

	/**
	 * @inheritDoc
	 */
	public function getIconClass(): string {
		return 'icon-openproject';
	}

	/**
	 * @inheritDoc
	 */
	public function getUrl(): ?string {
		return $this->url->linkToRoute('settings.PersonalSettings.index', ['section' => 'connected-accounts']);
	}

	/**
	 * @inheritDoc
	 */
	public function load(): void {
		Util::addScript(Application::APP_ID, Application::APP_ID . '-dashboard');
		Util::addStyle(Application::APP_ID, 'dashboard');

		$oauthConnectionResult = $this->config->getUserValue(
			$this->user->getUID(), Application::APP_ID, 'oauth_connection_result', ''
		);
		$this->config->deleteUserValue(
			$this->user->getUID(), Application::APP_ID, 'oauth_connection_result'
		);

		$authorizationMethod = $this->config->getAppValue(Application::APP_ID, 'authorization_method', '');
		$this->initialStateService->provideInitialState('authorization_method', $authorizationMethod);
		$this->initialStateService->provideInitialState(
			'admin_config_ok', OpenProjectAPIService::isAdminConfigOk($this->config)
		);

		$token = $this->openProjectAPIService->getAccessToken($this->user->getUID());
		$this->initialStateService->provideInitialState('user-has-oidc-token', boolval($token));
		$this->initialStateService->provideInitialState('oidc_user', $this->openProjectAPIService->isOIDCUser());

		// for 'oauth2' state to be loaded
		$this->initialStateService->provideInitialState(
			'oauth-connection-result', $oauthConnectionResult
		);
		$oauthConnectionErrorMessage = $this->config->getUserValue(
			$this->user->getUID(), Application::APP_ID, 'oauth_connection_error_message', ''
		);
		$this->config->deleteUserValue(
			$this->user->getUID(), Application::APP_ID, 'oauth_connection_error_message'
		);
		$this->initialStateService->provideInitialState(
			'oauth-connection-error-message', $oauthConnectionErrorMessage
		);
	}
}
