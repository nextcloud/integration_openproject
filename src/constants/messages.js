/**
 * SPDX-FileCopyrightText: 2025 Jankari Tech Pvt. Ltd.
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import { translate as t } from '@nextcloud/l10n'
import APP_ID from './appID.js'

export const messages = {
	installLatestVersionNow: t(APP_ID, 'Install latest version now'),
	downloadAndEnableApp: t(APP_ID, 'Download and enable it'),
	featureNotAvailable: t(APP_ID, 'This feature is not available for this user account'),
	opConnectionUnauthorized: t(APP_ID, 'Unauthorized to connect to OpenProject'),
	opClientId: t(APP_ID, 'OpenProject client ID'),
	tokenExchangeFormLabel: t(APP_ID, 'Token Exchange'),
	enableTokenExchange: t(APP_ID, 'Enable token exchange'),
	opClientIdHintText: t(APP_ID, 'You can get this value from your identity provider when you configure the client'),
	nextcloudHubProvider: t(APP_ID, 'Nextcloud Hub'),
	externalOIDCProvider: t(APP_ID, 'External Provider'),
	tokenExchangeHintText: t(APP_ID, 'When enabled, the app will try to obtain a token for the given audience from the identity provider. If disabled, it will use the access token obtained during the login process.'),
	opRequiredVersionAndPlanHint: t(APP_ID, 'Requires OpenProject version 16.0 (or higher) and an active Corporate plan.'),
}

export const messagesFmt = {
	appNotEnabledOrUnsupported: (app, version) => t(APP_ID, 'This feature requires version {version} (or higher) of "{app}" app. Please install or update the app.', { app, version }),
	configureOIDCProviders: (settingsLink) => t(APP_ID, 'You can configure OIDC providers in the {settingsLink}', { settingsLink }, null, { escape: false, sanitize: false }),
}
