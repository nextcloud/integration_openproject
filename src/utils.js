/**
 * SPDX-FileCopyrightText: 2022-2025 Jankari Tech Pvt. Ltd.
 * SPDX-FileCopyrightText: 2021 Nextcloud GmbH and Nextcloud contributors
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import { showError, showSuccess } from '@nextcloud/dialogs'
import { translate as t } from '@nextcloud/l10n'

let mytimer = 0
export function delay(callback, ms) {
	return function() {
		const context = this
		const args = arguments
		clearTimeout(mytimer)
		mytimer = setTimeout(function() {
			callback.apply(context, args)
		}, ms || 0)
	}
}

export function checkOauthConnectionResult(oauthConnectionResult, oauthConnectionErrorMessage) {
	if (oauthConnectionResult === 'success') {
		showSuccess(t('integration_openproject', 'Successfully connected to OpenProject!'))
	} else if (oauthConnectionResult === 'error') {
		showError(
			t(
				'integration_openproject',
				'OAuth access token could not be obtained:',
			) + ' ' + oauthConnectionErrorMessage,
		)
	}
}

export const STATE = {
	OK: 'ok',
	ERROR: 'error',
	LOADING: 'loading',
	NO_TOKEN: 'no-token',
	CONNECTION_ERROR: 'connection-error',
	FAILED_FETCHING_WORKPACKAGES: 'failed-fetching-workpackages',
}

export const F_MODES = {
	VIEW: 0,
	EDIT: 1,
	DISABLE: 2,
	NEW: 3,
}

export const FORM = {
	SERVER: 0,
	AUTHORIZATION_METHOD: 1,
	AUTHORIZATION_SETTING: 2,
	OP_OAUTH: 3,
	NC_OAUTH: 4,
	GROUP_FOLDER: 5,
	APP_PASSWORD: 6,
}

export const WORKPACKAGES_SEARCH_ORIGIN = {
	PROJECT_TAB: 'project-tab',
	LINK_MULTIPLE_FILES_MODAL: 'link-multiple-files-modal',
}
export const USER_SETTINGS = {
	NAVIGATION_LINK_DESCRIPTION: t('integration_openproject', 'Displays a link to your OpenProject instance in the Nextcloud header.'),
	UNIFIED_SEARCH_DESCRIPTION: t('integration_openproject', 'Allows you to search OpenProject work packages via the universal search bar in Nextcloud.'),
}

export const NO_OPTION_TEXT_STATE = {
	START_TYPING: 0,
	SEARCHING: 1,
	RESULT: 2,
}

export const AUTH_METHOD = {
	OAUTH2: 'oauth2',
	OIDC: 'oidc',
}

export const AUTH_METHOD_LABEL = {
	OAUTH2: t('integration_openproject', 'Two-way OAuth 2.0 authorization code flow'),
	OIDC: t('integration_openproject', 'Single-Sign-On through OpenID Connect Identity Provider'),
}

export const SSO_PROVIDER_TYPE = {
	nextcloudHub: 'nextcloud_hub',
	external: 'external',
}

export const SSO_PROVIDER_LABEL = {
	nextcloudHub: 'Nextcloud Hub',
}
