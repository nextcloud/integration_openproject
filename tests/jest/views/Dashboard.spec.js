/* jshint esversion: 8 */

/**
 * SPDX-FileCopyrightText: 2022-2025 Jankari Tech Pvt. Ltd.
 * SPDX-FileCopyrightText: 2023 Nextcloud GmbH and Nextcloud contributors
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import { shallowMount, createLocalVue } from '@vue/test-utils'
// eslint-disable-next-line n/no-unpublished-import
import flushPromises from 'flush-promises'
import util from 'util'
import { showError, showSuccess } from '@nextcloud/dialogs'
import { generateOcsUrl } from '@nextcloud/router'
import axios from '@nextcloud/axios'
import Dashboard from '../../../src/views/Dashboard.vue'
import { STATE, AUTH_METHOD, checkOauthConnectionResult } from '../../../src/utils.js'
import notificationsResponse from '../fixtures/notificationsResponse.json'
import { messages } from '../../../src/constants/messages.js'

jest.mock('@nextcloud/axios')
jest.mock('@nextcloud/dialogs', () => ({
	getLanguage: jest.fn(() => ''),
	showError: jest.fn(),
	showSuccess: jest.fn(),
}))
jest.mock('@nextcloud/initial-state', () => {
	const originalModule = jest.requireActual('@nextcloud/initial-state')
	return {
		__esModule: true,
		...originalModule,
		default: jest.fn(),
		loadState: jest.fn(() => ({ version: '32' })),
	}
})
jest.mock('../../../src/utils.js', () => ({
	...jest.requireActual('../../../src/utils.js'),
	checkOauthConnectionResult: jest.fn(),
}))

global.OCA = {}
global.OC = {}
const localVue = createLocalVue()

// url
const opUrl = generateOcsUrl('/apps/integration_openproject/api/v1/url')
const notificationUrl = generateOcsUrl('/apps/integration_openproject/api/v1/notifications')
const wpNotificationsUrl = generateOcsUrl('/apps/integration_openproject/api/v1/work-packages/%s/notifications')

describe('Dashboard.vue', () => {
	const errorLabelSelector = 'errorlabel-stub'
	const emptyContentSelector = 'emptycontent-stub'

	const defaultState = {
		authMethods: AUTH_METHOD,
	}

	let spyAxiosGet, spyLaunchLoop

	beforeEach(async () => {
		spyAxiosGet = jest.spyOn(axios, 'get')
			.mockImplementation(getAxiosGetMockFn())
		spyLaunchLoop = jest.spyOn(Dashboard.methods, 'launchLoop')
	})
	afterEach(() => {
		jest.clearAllMocks()
		jest.restoreAllMocks()
	})

	describe('auth method: OAUTH2', () => {
		const commonState = {
			...defaultState,
			authMethod: AUTH_METHOD.OAUTH2,
			oidc_user: false,
			userHasOidcToken: false,
		}
		describe('admin config is not ok', () => {
			it('should show empty content', async () => {
				const wrapper = getWrapper({
					oauthConnectionErrorMessage: '',
					oauthConnectionResult: '',
					isAdminConfigOk: false,
					...commonState,
				})
				await flushPromises()

				expect(wrapper.vm.state).toBe(STATE.ERROR)
				const emptyContent = wrapper.find(emptyContentSelector)
				expect(emptyContent.exists()).toBe(true)
				expect(emptyContent.attributes().state).toBe(STATE.ERROR)
				expect(emptyContent.attributes().authmethod).toBe(AUTH_METHOD.OAUTH2)
				expect(emptyContent.attributes().dashboard).toBe('true')
				expect(emptyContent.attributes().isadminconfigok).toBeFalsy()

				expect(wrapper.find(errorLabelSelector).exists()).toBe(false)
				expect(spyLaunchLoop).toHaveBeenCalledTimes(1)
				expect(spyAxiosGet).not.toBeCalled()
				expect(checkOauthConnectionResult).not.toBeCalled()
			})
		})

		describe('admin config is ok', () => {
			describe('not connected to OP', () => {
				beforeEach(async () => {
					spyAxiosGet.mockRestore()
					spyAxiosGet = jest.spyOn(axios, 'get')
						// eslint-disable-next-line prefer-promise-reject-errors
						.mockImplementation(getAxiosGetMockFn(Promise.reject({ response: { status: 401 } })))
				})
				it('should show empty content', async () => {
					const wrapper = getWrapper({
						oauthConnectionErrorMessage: '',
						oauthConnectionResult: '',
						isAdminConfigOk: true,
						...commonState,
					})
					await flushPromises()

					expect(wrapper.vm.state).toBe(STATE.NO_TOKEN)
					const emptyContent = wrapper.find(emptyContentSelector)
					expect(emptyContent.exists()).toBe(true)
					expect(emptyContent.attributes().state).toBe(STATE.NO_TOKEN)
					expect(emptyContent.attributes().authmethod).toBe(AUTH_METHOD.OAUTH2)
					expect(emptyContent.attributes().dashboard).toBe('true')
					expect(emptyContent.attributes().isadminconfigok).toBe('true')

					expect(wrapper.find(errorLabelSelector).exists()).toBe(false)
					expect(spyAxiosGet).toBeCalledWith(opUrl)
					expect(spyAxiosGet).toBeCalledWith(notificationUrl)
					expect(checkOauthConnectionResult).toHaveBeenCalledTimes(1)
				})
			})

			describe('connected to OP', () => {
				it('should show the notification items', async () => {
					const wrapper = getWrapper({
						oauthConnectionErrorMessage: 'some-token',
						oauthConnectionResult: 'success',
						isAdminConfigOk: true,
						...commonState,
					})
					await flushPromises()

					expect(spyAxiosGet).toBeCalledWith(opUrl)
					expect(spyAxiosGet).toBeCalledWith(notificationUrl)
					expect(wrapper.vm.state).toBe(STATE.OK)
					expect(checkOauthConnectionResult).toHaveBeenCalledTimes(1)
					expect(wrapper.vm.items).toMatchSnapshot()
				})
			})
		})
	})

	describe('auth method: OIDC', () => {
		const commonState = {
			...defaultState,
			oauthConnectionErrorMessage: '',
			oauthConnectionResult: '',
			authMethod: AUTH_METHOD.OIDC,
		}

		describe('OIDC user', () => {
			const localState = { ...commonState, oidc_user: true }
			describe('admin config is not ok', () => {
				it('should show empty content', async () => {
					const wrapper = getWrapper({
						isAdminConfigOk: false,
						...localState,
					})
					await flushPromises()

					expect(wrapper.vm.state).toBe(STATE.ERROR)
					const emptyContent = wrapper.find(emptyContentSelector)
					expect(emptyContent.exists()).toBe(true)
					expect(emptyContent.attributes().state).toBe(STATE.ERROR)
					expect(emptyContent.attributes().authmethod).toBe(AUTH_METHOD.OIDC)
					expect(emptyContent.attributes().dashboard).toBe('true')
					expect(emptyContent.attributes().isadminconfigok).toBeFalsy()

					expect(wrapper.find(errorLabelSelector).exists()).toBe(false)
					expect(spyLaunchLoop).toHaveBeenCalledTimes(1)
					expect(spyAxiosGet).not.toBeCalled()
					expect(checkOauthConnectionResult).not.toBeCalled()
				})
			})

			describe('admin config is ok', () => {
				describe('not connected to OP', () => {
					beforeEach(async () => {
						spyAxiosGet.mockRestore()
						spyAxiosGet = jest.spyOn(axios, 'get')
							// eslint-disable-next-line prefer-promise-reject-errors
							.mockImplementation(getAxiosGetMockFn(Promise.reject({ response: { status: 401 } })))
					})
					it('should show unauthorized error', async () => {
						const wrapper = getWrapper({
							...localState,
							isAdminConfigOk: true,
							userHasOidcToken: false,
						})
						await flushPromises()

						expect(wrapper.vm.state).toBe(STATE.NO_TOKEN)
						expect(wrapper.find(errorLabelSelector).attributes().error).toBe(messages.opConnectionUnauthorized)

						expect(spyLaunchLoop).toHaveBeenCalledTimes(1)
						expect(wrapper.find(emptyContentSelector).exists()).toBe(false)
						expect(spyAxiosGet).toBeCalledWith(opUrl)
						expect(spyAxiosGet).toBeCalledWith(notificationUrl)
						expect(checkOauthConnectionResult).not.toBeCalled()
					})
				})

				describe('connected to OP', () => {
					describe('invalid token', () => {
						beforeEach(async () => {
							spyAxiosGet.mockRestore()
							spyAxiosGet = jest.spyOn(axios, 'get')
							// eslint-disable-next-line prefer-promise-reject-errors
								.mockImplementation(getAxiosGetMockFn(Promise.reject({ response: { status: 401 } })))
						})
						it('should show unauthorized error if token is not valid', async () => {
							const wrapper = getWrapper({
								...localState,
								isAdminConfigOk: true,
								userHasOidcToken: true,
							})
							await flushPromises()

							expect(wrapper.vm.state).toBe(STATE.NO_TOKEN)
							expect(wrapper.find(errorLabelSelector).attributes().error).toBe(messages.opConnectionUnauthorized)

							expect(spyLaunchLoop).toHaveBeenCalledTimes(1)
							expect(wrapper.find(emptyContentSelector).exists()).toBe(false)
							expect(spyAxiosGet).toBeCalledWith(opUrl)
							expect(spyAxiosGet).toBeCalledWith(notificationUrl)
							expect(checkOauthConnectionResult).not.toBeCalled()
						})
					})

					it('should show the notification items', async () => {
						const wrapper = getWrapper({
							...localState,
							isAdminConfigOk: true,
							userHasOidcToken: true,
						})
						await flushPromises()

						expect(spyAxiosGet).toBeCalledWith(opUrl)
						expect(spyAxiosGet).toBeCalledWith(notificationUrl)
						expect(wrapper.vm.state).toBe(STATE.OK)
						expect(checkOauthConnectionResult).not.toBeCalled()
						expect(wrapper.vm.items).toMatchSnapshot()
					})
				})
			})
		})

		describe('non OIDC user', () => {
			const localState = { ...commonState, userHasOidcToken: false, oidc_user: false }
			describe('admin config is not ok', () => {
				it('should show empty content', async () => {
					const wrapper = getWrapper({
						isAdminConfigOk: false,
						...localState,
					})
					await flushPromises()

					expect(wrapper.vm.state).toBe(STATE.ERROR)
					const emptyContent = wrapper.find(emptyContentSelector)
					expect(emptyContent.exists()).toBe(true)
					expect(emptyContent.attributes().state).toBe(STATE.ERROR)
					expect(emptyContent.attributes().authmethod).toBe(AUTH_METHOD.OIDC)
					expect(emptyContent.attributes().dashboard).toBe('true')
					expect(emptyContent.attributes().isadminconfigok).toBeFalsy()

					expect(wrapper.find(errorLabelSelector).exists()).toBe(false)
					expect(spyLaunchLoop).toHaveBeenCalledTimes(1)
					expect(spyAxiosGet).not.toBeCalled()
					expect(checkOauthConnectionResult).not.toBeCalled()
				})
			})

			describe('admin config is ok', () => {
				it('should show feature not available error', async () => {
					const wrapper = getWrapper({
						isAdminConfigOk: true,
						...localState,
					})
					await flushPromises()

					expect(wrapper.find(errorLabelSelector).attributes().error).toBe(messages.featureNotAvailable)

					expect(spyLaunchLoop).toHaveBeenCalledTimes(1)
					expect(wrapper.find(emptyContentSelector).exists()).toBe(false)
					expect(checkOauthConnectionResult).not.toBeCalled()
					expect(spyAxiosGet).not.toBeCalled()
				})
			})
		})
	})

	describe('mark as read', () => {
		const commonState = {
			...defaultState,
			authMethod: AUTH_METHOD.OAUTH2,
			oidc_user: false,
			userHasOidcToken: false,
		}
		const notification = { id: 1 }

		let spyAxiosDelete, spyFetchNotifications
		beforeEach(async () => {
			spyAxiosDelete = jest.spyOn(axios, 'delete')
				.mockImplementation(() => Promise.resolve({}))
			spyFetchNotifications = jest.spyOn(Dashboard.methods, 'fetchNotifications')
		})
		afterEach(() => {
			jest.clearAllMocks()
			jest.restoreAllMocks()
		})

		it('should re-fetch notifications on success', async () => {
			const wrapper = getWrapper({
				oauthConnectionErrorMessage: 'some-token',
				oauthConnectionResult: 'success',
				isAdminConfigOk: true,
				...commonState,
			})
			await flushPromises()
			wrapper.vm.onMarkAsRead(notification)
			await flushPromises()

			expect(spyAxiosDelete).toBeCalledWith(util.format(wpNotificationsUrl, notification.id))
			expect(spyFetchNotifications).toHaveBeenCalledTimes(2)
			expect(showSuccess).toHaveBeenCalledTimes(1)
			expect(showError).toHaveBeenCalledTimes(0)
		})
		it('should show error toast message on failure', async () => {
			spyAxiosDelete = jest.spyOn(axios, 'delete')
			// eslint-disable-next-line prefer-promise-reject-errors
				.mockImplementation(() => Promise.reject('mark as read failed'))
			const wrapper = getWrapper({
				oauthConnectionErrorMessage: 'some-token',
				oauthConnectionResult: 'success',
				isAdminConfigOk: true,
				...commonState,
			})
			await flushPromises()
			wrapper.vm.onMarkAsRead(notification)
			await flushPromises()

			expect(spyAxiosDelete).toBeCalledWith(util.format(wpNotificationsUrl, notification.id))
			expect(spyFetchNotifications).toHaveBeenCalledTimes(1)
			expect(showError).toHaveBeenCalledTimes(1)
			expect(showSuccess).toHaveBeenCalledTimes(0)
		})
	})
})

function getAxiosGetMockFn(notificationResponse) {
	if (!notificationResponse) {
		notificationResponse = Promise.resolve({ data: { ocs: { data: notificationsResponse } } })
	}
	return (url) => {
		switch (url) {
		case opUrl:
			return Promise.resolve({ data: { ocs: { data: 'http://openproject.org' } } })
		case notificationUrl:
			return notificationResponse
		default:
			return Promise.reject(new Error('unexpected url'))
		}
	}
}

function getWrapper(data = {}) {
	return shallowMount(
		Dashboard,
		{
			localVue,
			mocks: {
				t: (app, msg) => msg,
			},
			propsData: {
				title: 'dashboard',
			},
			data: () => data,
		})
}
