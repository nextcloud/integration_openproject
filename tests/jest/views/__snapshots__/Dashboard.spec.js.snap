// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard.vue auth method: OAUTH2 admin config is ok connected to OP should show the notification items 1`] = `
[
  {
    "avatarUrl": "http://localhost/apps/integration_openproject/api/v1/avatar?userId=8&userName=Admin%20de%20DEV%20user",
    "avatarUsername": "Admin de DEV userz",
    "id": "17",
    "mainText": "(6) Create wireframes for new landing page",
    "overlayIconUrl": "",
    "subText": "Scrum project - accountable, commented, mentioned, Date alert",
    "targetUrl": "http://openproject.org/notifications/details/17/activity/",
  },
  {
    "avatarUrl": "http://localhost/apps/integration_openproject/api/v1/avatar?userId=9&userName=Artur%20Neumann",
    "avatarUsername": "Artur Neumannz",
    "id": "18",
    "mainText": "(1) Contact form",
    "overlayIconUrl": "",
    "subText": "Scrum project - mentioned",
    "targetUrl": "http://openproject.org/notifications/details/18/activity/",
  },
  {
    "avatarUrl": "http://localhost/apps/integration_openproject/api/v1/avatar?userName=",
    "avatarUsername": "undefinedz",
    "id": "24",
    "mainText": "(1) Choose a content management system",
    "overlayIconUrl": "",
    "subText": "Demo project - Date alert",
    "targetUrl": "http://openproject.org/notifications/details/24/activity/",
  },
  {
    "avatarUrl": "http://localhost/apps/integration_openproject/api/v1/avatar?userId=8&userName=Admin%20de%20DEV%20user",
    "avatarUsername": "Admin de DEV userz",
    "id": "36",
    "mainText": "(2) write a software",
    "overlayIconUrl": "",
    "subText": "Dev-large - assignee, watcher",
    "targetUrl": "http://openproject.org/notifications/details/36/activity/",
  },
]
`;

exports[`Dashboard.vue auth method: OIDC OIDC user admin config is ok connected to OP should show the notification items 1`] = `
[
  {
    "avatarUrl": "http://localhost/apps/integration_openproject/api/v1/avatar?userId=8&userName=Admin%20de%20DEV%20user",
    "avatarUsername": "Admin de DEV userz",
    "id": "17",
    "mainText": "(6) Create wireframes for new landing page",
    "overlayIconUrl": "",
    "subText": "Scrum project - accountable, commented, mentioned, Date alert",
    "targetUrl": "http://openproject.org/notifications/details/17/activity/",
  },
  {
    "avatarUrl": "http://localhost/apps/integration_openproject/api/v1/avatar?userId=9&userName=Artur%20Neumann",
    "avatarUsername": "Artur Neumannz",
    "id": "18",
    "mainText": "(1) Contact form",
    "overlayIconUrl": "",
    "subText": "Scrum project - mentioned",
    "targetUrl": "http://openproject.org/notifications/details/18/activity/",
  },
  {
    "avatarUrl": "http://localhost/apps/integration_openproject/api/v1/avatar?userName=",
    "avatarUsername": "undefinedz",
    "id": "24",
    "mainText": "(1) Choose a content management system",
    "overlayIconUrl": "",
    "subText": "Demo project - Date alert",
    "targetUrl": "http://openproject.org/notifications/details/24/activity/",
  },
  {
    "avatarUrl": "http://localhost/apps/integration_openproject/api/v1/avatar?userId=8&userName=Admin%20de%20DEV%20user",
    "avatarUsername": "Admin de DEV userz",
    "id": "36",
    "mainText": "(2) write a software",
    "overlayIconUrl": "",
    "subText": "Dev-large - assignee, watcher",
    "targetUrl": "http://openproject.org/notifications/details/36/activity/",
  },
]
`;
